node: create podresources endpoint also on windows

ffromani opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
re-implement LocalEndpoint on windows, prerequisite to enable the podresources API on windows and thus to graduate the feature to GA

#### Which issue(s) this PR fixes:
Needed for https://github.com/kubernetes/enhancements/issues/3743

#### Special notes for your reviewer:
I'm following the directions outlined in https://github.com/kubernetes/kubernetes/issues/78628 about the naming and the layout of the URL

#### Does this PR introduce a user-facing change?
```release-note
Enables the node-local kubelet podresources API endpoint on windows, alongside unix.
```

ffromani committed:
node: re-implement Localendpoint on windows

this will allows us to move forward with the podresources
endpoint GA graduation.

xref: https://github.com/kubernetes/kubernetes/issues/78628

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
/retest

ffromani committed:
gce: podresources:  don't disable on windows

Now that podresources socket has been fixed on windows,
we can re-enable on GCE CI.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
@marosset hi! could you please suggest reviewers? thanks a ton!

marosset commented:
/assign @claudiubelu 
Can you please review this?

I'll also take a look

ffromani commented:
/retest
it seems infra issues, changes since the last green run are unlikely to cause breakage (and they are covered by a unit test which is still green)

ffromani commented:
/retest
(see: https://github.com/kubernetes/kubernetes/pull/115133#issuecomment-1403279176)

ffromani commented:
```
 I0125 12:49:11.539] Call:  gcloud auth activate-service-account --key-file=/etc/service-account/service-account.json
E0125 12:49:12.419] unexpected error
Traceback (most recent call last):
  File "./test-infra/jenkins/bootstrap.py", line 1094, in bootstrap
    setup_credentials(call, args.service_account, upload)
  File "./test-infra/jenkins/bootstrap.py", line 772, in setup_credentials
    '--key-file=%s' % os.environ[SERVICE_ACCOUNT_ENV],
  File "./test-infra/jenkins/bootstrap.py", line 1054, in <lambda>
    call = lambda *a, **kw: _call(end, *a, **kw)
  File "./test-infra/jenkins/bootstrap.py", line 132, in _call
    if read_all(end, *reads[fdesc]):
  File "./test-infra/jenkins/bootstrap.py", line 68, in read_all
    append(line.rstrip('\n'))
TypeError: a bytes-like object is required, not 'str'
I0125 12:49:12.420] Call:  gcloud auth activate-service-account --key-file=/etc/service-account/service-account.json
Traceback (most recent call last):
  File "./test-infra/jenkins/bootstrap.py", line 1202, in <module>
    bootstrap(ARGS)
  File "./test-infra/jenkins/bootstrap.py", line 1117, in bootstrap
    setup_credentials(call, args.service_account, upload)
  File "./test-infra/jenkins/bootstrap.py", line 772, in setup_credentials
    '--key-file=%s' % os.environ[SERVICE_ACCOUNT_ENV],
  File "./test-infra/jenkins/bootstrap.py", line 1054, in <lambda>
    call = lambda *a, **kw: _call(end, *a, **kw)
  File "./test-infra/jenkins/bootstrap.py", line 132, in _call
    if read_all(end, *reads[fdesc]):
  File "./test-infra/jenkins/bootstrap.py", line 68, in read_all
    append(line.rstrip('\n'))
TypeError: a bytes-like object is required, not 'str' 
```

claudiubelu commented:
/test pull-ci-kubernetes-unit-windows

ffromani commented:
/test pull-kubernetes-e2e-capz-windows-containerd

bart0sh commented:
/triage accepted
/priority important-longterm

bart0sh commented:
/retest

claudiubelu commented:
It seems that the unit test introduced here is failing on Windows:

```
{Failed  === RUN   TestLocalEndpoint
    util_windows_test.go:344: 
        	Error Trace:	C:\kubernetes\pkg\kubelet\util\util_windows_test.go:344
        	Error:      	Not equal: 
        	            	expected: "npipe://.pipe/kubelet/pod-resources"
        	            	actual  : "npipe://./pipe/kubelet/pod-resources"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-npipe://.pipe/kubelet/pod-resources
        	            	+npipe://./pipe/kubelet/pod-resources
        	Test:       	TestLocalEndpoint
--- FAIL: TestLocalEndpoint (0.00s)
}
```

ffromani commented:
/test pull-ci-kubernetes-unit-windows

ffromani commented:
/test pull-kubernetes-e2e-capz-windows-containerd

ffromani commented:
> It seems that the unit test introduced here is failing on Windows:
> 
> ```
> {Failed  === RUN   TestLocalEndpoint
>     util_windows_test.go:344: 
>         	Error Trace:	C:\kubernetes\pkg\kubelet\util\util_windows_test.go:344
>         	Error:      	Not equal: 
>         	            	expected: "npipe://.pipe/kubelet/pod-resources"
>         	            	actual  : "npipe://./pipe/kubelet/pod-resources"
>         	            	
>         	            	Diff:
>         	            	--- Expected
>         	            	+++ Actual
>         	            	@@ -1 +1 @@
>         	            	-npipe://.pipe/kubelet/pod-resources
>         	            	+npipe://./pipe/kubelet/pod-resources
>         	Test:       	TestLocalEndpoint
> --- FAIL: TestLocalEndpoint (0.00s)
> }
> ```

thanks, fixed with the last upload. But I'm not completely sure this is a good path layout. I'm inexperienced on windows, trying to follow up advice found in https://github.com/kubernetes/kubernetes/issues/78628

claudiubelu commented:
> > > > It seems that the unit test introduced here is failing on Windows:
> > > > ```
> > > > {Failed  === RUN   TestLocalEndpoint
> > > >     util_windows_test.go:344: 
> > > >         	Error Trace:	C:\kubernetes\pkg\kubelet\util\util_windows_test.go:344
> > > >         	Error:      	Not equal: 
> > > >         	            	expected: "npipe://.pipe/kubelet/pod-resources"
> > > >         	            	actual  : "npipe://./pipe/kubelet/pod-resources"
> > > >         	            	
> > > >         	            	Diff:
> > > >         	            	--- Expected
> > > >         	            	+++ Actual
> > > >         	            	@@ -1 +1 @@
> > > >         	            	-npipe://.pipe/kubelet/pod-resources
> > > >         	            	+npipe://./pipe/kubelet/pod-resources
> > > >         	Test:       	TestLocalEndpoint
> > > > --- FAIL: TestLocalEndpoint (0.00s)
> > > > }
> > > > ```
> > > 
> > > 
> > > thanks, fixed with the last upload. But I'm not completely sure this is a good path layout. I'm inexperienced on windows, trying to follow up advice found in #78628
> > 
> > 
> > Hmm, right. It seems that the documentation specifies that named pipes should have backslashes instead of forward slashes [1][2].
> > [1] https://learn.microsoft.com/en-us/windows/win32/ipc/pipe-names [2] https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createnamedpipea#parameters
> 
> thanks for the link to the docs, this helps. I will experiment with forward slash. Do we need to have the change merged to exercise e2e tests on CI or the e2e CI lane will pick the new code I'm adding here?
> 
> From sig-node perspective this work is likely to slip to 1.28 so we have some room to monitor CI and be confident this actually works. I'm committed to fix windows support to be able to graduate podresources API for good (xref: [c2372f3](https://github.com/kubernetes/kubernetes/commit/c2372f3c8318cb178c7e5c10005e379754cc43a0)). The "only" caveat is I'm very inexperienced on windows so please bear with me :)

Hmm, you can simply trigger the Windows capz jobs by commenting:

/test pull-kubernetes-e2e-capz-windows-containerd

I see that the feature is behind a specific feature gate, ``KubeletPodResources``, which is enabled by default. 

Looking at the capz job that ran here [1], I don't see the feature gate being turned off in a Windows node's kubelet logs [2]. So, this means that you could test ``KubeletPodResources`` as mentioned above. Now, the question is, are there are E2E tests for it? We're currently only running ``[Conformance]`` and ``[NodeConformance]`` tests, but a quick way to simply test the feature on Windows would be to temprarily add the ``[Conformance]`` tag to the E2E tests, trigger the jobs as above and see it pass.

[1] https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/115133/pull-kubernetes-e2e-capz-windows-containerd/1620823317965443072
[2] https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/115133/pull-kubernetes-e2e-capz-windows-containerd/1620823317965443072/artifacts/clusters/capz-conf-vmhcgc/machines/capz-conf-vmhcgc-md-win-cc848b8c9-zqmgg/kubelet.log

ffromani commented:
going back to WIP because I need to revisit the path constructions per windows documentation

ffromani commented:
/test pull-ci-kubernetes-unit-windows

ffromani commented:
As part of this work, I'll expand the podresources e2e tests we have to work also on windows. The current testsuite is has a lot of implicit or explicit dependencies on linux (paths, command line tools...), so needs some improvements. Later we will check if we can promote these tests to `Conformance`/`NodeConformance`

ffromani commented:
/test pull-ci-kubernetes-unit-windows

ffromani commented:
/test pull-kubernetes-e2e-capz-windows-containerd

ffromani commented:
```
 Conformance Tests [It] conformance-tests
/home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/conformance_test.go:100
  [FAILED] Timed out after 1500.000s.
  Timed out waiting for 2 nodes to be created for MachineDeployment capz-conf-rdkg21/capz-conf-rdkg21-md-win
  Expected
      <int>: 0
  to equal
      <int>: 2
  In [It] at: /home/prow/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.3.2/framework/machinedeployment_helpers.go:131 @ 02/05/23 08:20:53.161 
```
failure still unrelated to the matter here

ffromani commented:
/test pull-kubernetes-e2e-capz-windows-containerd

ffromani commented:
/test pull-kubernetes-e2e-capz-windows-containerd

ffromani commented:
/test pull-kubernetes-e2e-capz-windows-containerd
last attempt. It seems infra issue to me, the lane fails well before starting actual tests. If still fails, I'll ask around in sig-windows

ffromani commented:
/test pull-kubernetes-e2e-capz-windows-containerd

ffromani commented:
rebased

ffromani commented:
/test pull-kubernetes-e2e-capz-windows-containerd

ffromani commented:
/test pull-kubernetes-e2e-capz-windows-containerd

ffromani commented:
/retest

```
k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/test/integration: TestChangeCRD
```

ffromani commented:
```
k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/test/integration: TestChangeCRD
```
this more and more looks like a flake now

ffromani commented:
/retest

```
Failed  === RUN   Test_Run_OneVolumeDetachFailNodeWithReadWriteOnce
```

ffromani commented:
The PR is now reviewable again. I'm using sig-windows-dev-tools to do the final validation.

From the latest run: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/115133/pull-kubernetes-e2e-capz-windows-containerd/1633479919428177920/artifacts/clusters/capz-conf-k7qi1k/machines/capz-conf-k7qi1k-md-win-b4cd58679-x5cwd/kubelet.log

we can see
```
I0308 15:41:09.794567    3552 server.go:223] "Starting to serve the podresources API" socket="npipe://\\\\.\\pipe\\kubelet\\pod-resources"
```

ffromani commented:
did extra manual validation using sig-windows-dev-tools and https://github.com/k8stopologyawareschedwg/podresourcesapi-tools:
```
PS C:\k> ..\client.exe -podresources-socket npipe://\\\\.\\pipe\\kubelet\\pod-resources list
{"pod_resources":[{"name":"calico-node-windows-zssp5","namespace":"calico-system","containers":[{"name":"node"},{"name":"felix"}]}]}
```

ffromani commented:
> did extra manual validation using sig-windows-dev-tools and https://github.com/k8stopologyawareschedwg/podresourcesapi-tools:
> 
> ```
> PS C:\k> ..\client.exe -podresources-socket npipe://\\\\.\\pipe\\kubelet\\pod-resources list
> {"pod_resources":[{"name":"calico-node-windows-zssp5","namespace":"calico-system","containers":[{"name":"node"},{"name":"felix"}]}]}
> ```

There's an ongoing effort by myself and others to promote some of the existing podresources e2e tests to nodeconformance, so the windows lane will always run them. We will work on this.

jsturtevant commented on pkg/kubelet/server/server.go:
```diff
@@ -220,6 +220,7 @@ func ListenAndServePodResources(socket string, podsProvider podresources.PodsPro
 	}
 
+	klog.InfoS("Starting to serve the podresources API", "socket", socket)
 	if err := server.Serve(l); err != nil {
```
Is this creating a Unix Domain socket or named pipe?

ffromani replied:
both: it's a unix domain socket on unix[1] and a named pipe on windows. Maybe should I rename the to read
```
klog.InfoS("Starting to serve the podresources API", "endpoint", endpoint)
```
(of course rename the variable not just the log)

[1]: ```
root@controlplane:/home/vagrant# file /var/lib/kubelet/pod-resources/kubelet.sock 
/var/lib/kubelet/pod-resources/kubelet.sock: socket
```

jsturtevant replied:
Oh I see now, `pkg/kubelet/util/util_windows.go` has the logic to create the namepipe server.  Thanks!

jsturtevant commented on pkg/kubelet/util/util_windows_test.go:
```diff
@@ -317,3 +317,48 @@ func TestNormalizePath(t *testing.T) {
+			path:             "/var/lib/kubelet/pod-resources",
+			file:             "kubelet",
+			expectError:      false,
+			expectedFullPath: `npipe://\\.\pipe\kubelet\pod-resources`,
```
I don't think this is techinically wrong but we typically just specify it like the following: https://github.com/containerd/containerd/blob/71a7234f67feec950bbc25b9d89983d80d72a743/defaults/defaults_windows.go#L38

ffromani replied:
I see no problem changing the format. Would something like `\\.\pipe\kubelet-pod-resources` acceptable?

jsturtevant replied:
yea, I think that would make it consistent with alot of other tools

ffromani replied:
changed, thanks!

jsturtevant replied:
this looks the same?

jsturtevant replied:
@ffromani oh, I see, I meant `npipe://` isn't really needed.  and would remove some of the duplicate slashes 🤷

ffromani replied:
thanks for pointing this out. I can remove the `npipe://`. Which slashes are duplicate? according to the spec it seems to me that for example the leading `\\` is expected

ffromani replied:
@jsturtevant there's something I'm not sure about. On unix, the function returns the path prefixed by `unix://` - so it looks like an URL more than a naked path. And indeed the code expects the protocol part (see https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/util/util_windows.go#L92) so it seems to me we should do the same here? Perhaps adding forward AND back slash leads to some confusion, but I'm not sure what we should use here. I do see the kubelet using paths with only slashes:
```
I0310 14:25:08.155586    3988 flags.go:64] FLAG: --container-runtime-endpoint="npipe:////./pipe/containerd-containerd"
```
(from https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/115133/pull-kubernetes-e2e-capz-windows-containerd/1634188079650574336/artifacts/clusters/capz-conf-jq6aid/machines/capz-conf-jq6aid-md-win-6d7c774d58-klznw/kubelet.log)

ffromani committed:
kubelet: podresources: rename variable

on unix, the podresources endpoint is a unix domain socket;
on windows, the podresources endpoint is a named pipe;
rename the variables to convey this fact. No changes in behavior.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
/retest

the test seems failed for reasons unrelated to this PR, and the endpoint looks good:
```
I0308 19:28:32.744641    4720 server.go:223] "Starting to serve the podresources API" endpoint="npipe://\\\\.\\pipe\\kubelet-pod-resources"
```

ffromani commented:
/retest

ffromani commented:
/retest

it seems the lane failed in the cleanup stage, the tests actually passed

ffromani commented:
```
Failed  === RUN   TestSyncKnownPods
```
data race detected (!!)

ffromani commented:
/retest

SergeyKanzhelev commented on pkg/kubelet/util/util_windows.go:
```diff
@@ -117,9 +118,27 @@ func parseEndpoint(endpoint string) (string, string, error) {
+		// the user supplied root directory named like "pod-resources" or so.
+		return "", fmt.Errorf("cannot infer the podresources directory from path %q", path)
+	}
+	// windows pipes are expected to use forward slashes: https://learn.microsoft.com/en-us/windows/win32/ipc/pipe-names
```
nit:

```suggestion
	// windows pipes are expected to use forward slashes: https://learn.microsoft.com/windows/win32/ipc/pipe-names
```

ffromani replied:
thanks, fixed

SergeyKanzhelev reviewed (commented):
as long as this is good with sig windows, code looks good for me.

ffromani commented:
/retest

jsturtevant commented:
/retest 
`Pod can not be created: create pod test-pod` there were prow issues yesterday

ffromani commented:
rebased

jsturtevant commented:
/lgtm

ffromani commented:
/assign @yujuhong

marosset commented:
/lgtm

SergeyKanzhelev reviewed (approved):
/lgtm

ffromani commented:
/assing @dchen1107 @derekwaynecarr for the kubelet.go changes

SergeyKanzhelev commented:
> Does this PR introduce a user-facing change?

@ffromani perhaps this deserves some release notes

dchen1107 commented:
/lgtm
/approve

SergeyKanzhelev commented:
@ibabou please approve

/assign @ibabou

ffromani commented:
> > Does this PR introduce a user-facing change?
> 
> @ffromani perhaps this deserves some release notes

Added, thanks for the reminder.

ffromani commented:
thanks @SergeyKanzhelev @dchen1107 

rebased to fix conflicts, hopefully the last time for today :)

SergeyKanzhelev reviewed (approved):
/lgtm

re-applying

ibabou reviewed (approved):
Thanks Fransisco! LGTM

ibabou commented:
/lgtm

SergeyKanzhelev commented:
/retest-required

marosset commented:
/hold - missed code freeze deadline
/milestone clear

ffromani commented:
/retest

ffromani commented:
Aiming for early merge once the main branch gates open again. So I can monitor CI and wrap it up for 1.28

ffromani commented:
/hold cancel

we are gated by the milestone requirement (which this PR does not have and should not have) anyway
